source/extensions/filters/http/custom_response/custom_response_filter.cc

```diff
@@ -43,15 +44,29 @@ Http::FilterHeadersStatus CustomResponseFilter::encodeHeaders(Http::ResponseHead
     return filter_state->policy->encodeHeaders(headers, end_stream, *this);
   }
 
-  // Check for route specific config.
-  const FilterConfig* per_route_settings =
-      Http::Utility::resolveMostSpecificPerFilterConfig<FilterConfig>(decoder_callbacks_);
-  const FilterConfig* config_to_use = per_route_settings
-                                          ? static_cast<const FilterConfig*>(per_route_settings)
-                                          : static_cast<const FilterConfig*>(config_.get());
-  // Check if any custom response policy applies to this response.
-  const PolicySharedPtr policy =
-      config_to_use->getPolicy(headers, encoder_callbacks_->streamInfo());
+  PolicySharedPtr policy;
+  // Traverse up route typed per filter hierarchy till we find a matching
+  // policy.
+  bool match_found = false;
+  decoder_callbacks_->traversePerFilterConfig(
```

Comment:
This may be a wrong traversal IIUC (virtual-host will take over route, for example).

Reply:
I think it needs a test that validates that per route config overrides vhost config. I.e. then both take affect for the same status but translate it to different response codes.

Reply:
I had it wrong. Fixed it and added a test. Unfortunately this means we cannot exit early when a match is found, and will always check all configs.